src/widgets/gl_area.rs

```diff
@@ -0,0 +1,212 @@
+/// Drawing with GtkGLArea
+///
+/// The simplest way to draw using OpenGL commands in a GtkGLArea is to create a widget instance and connect to the "render" signal:
+///
+/// ```Rust,ignore
+/// // create a GtkGLArea instance
+/// GtkWidget *gl_area = gtk_gl_area_new ();
+///
+/// // connect to the "render" signal
+/// g_signal_connect (gl_area, "render", G_CALLBACK (render), NULL);
+/// ```
+///
+/// The render() function will be called when the GtkGLArea is ready for you to draw its content:
+///
+/// ```Rust,ignore
+/// fn render (area: &gtk::GLArea, context: &gdk::GLContext) -> bool {
+///     // inside this function it's safe to use GL; the given
+///     // #GdkGLContext has been made current to the drawable
+///     // surface used by the #GtkGLArea and the viewport has
+///     // already been set to be the size of the allocation
+/// 
+///     // we can start by clearing the buffer
+///     glClearColor(0, 0, 0, 0);
+///     glClear(GL_COLOR_BUFFER_BIT);
+///
+///     // draw your object
+///     draw_an_object();
+///
+///     // we completed our drawing; the draw commands will be
+///     // flushed at the end of the signal emission chain, and
+///     // the buffers will be drawn on the window
+///     true
+/// }
+/// ```
+///
+/// If you need to initialize OpenGL state, e.g. buffer objects or shaders, you should use the "realize" signal; you can use the "unrealize" signal to clean up. Since the GdkGLContext creation and initialization may fail, you will need to check for errors, using gtk_gl_area_get_error(). An example of how to safely initialize the GL state is:
+///
+/// ```Rust,ignore
+/// fn on_realize (area: &GLarea) {
+///     // We need to make the context current if we want to
+///     // call GL API
+///     area.make_current();
+///
+///     // If there were errors during the initialization or
+///     // when trying to make the context current, this
+///     // function will return a #GError for you to catch
+///     if area.get_error.is_some() {
+///         return;
+///     }
+///
+///     // You can also use gtk_gl_area_set_error() in order
+///     // to show eventual initialization errors on the
+///     // GtkGLArea widget itself
+///     let internal_error = NULL;
+///     match init_buffer_objects() {
+///         Some(e) => {
+///             area.set_error(e);
+///             return;
+///         }
+///         None => {}
+///     }
+///
+///     match init_shaders() {
+///         Some(e) => {
+///             area.set_error(e);
+///             return;
+///         }
+///         None => {}
+///     }
+/// }
+/// ```
+///
+/// If you need to change the options for creating the GdkGLContext you should use the "create-context" signal.
+struct_Widget!(GLArea);
+
+impl GLArea {
+    /// Creates a new GLArea widget.
+    pub fn new() -> Option<GLArea> {
```

Comment:
But the code should be able to continue to run anyway. Forcing this behavior isn't a good idea for me. If you don't have enough memory to allocate a new one, the previous ones still exist and work. Panicking because of this isn't a good idea !